signature/src/balrog.rs

```diff
@@ -0,0 +1,363 @@
+        let mut s = s.to_vec();
+
+        /* ASN.1 integers are signed, but ECDSA signature points are always unsigned.
+         * If the leading bit is zero, we must append an extra leading zero byte to
+         * ensure the sign is positive.
+         */
+        if r[0] & 0x80 != 0 {
+            r.insert(0, 0x00);
+        }
+        if s[0] & 0x80 != 0 {
+            s.insert(0, 0x00);
+        }
+
+        /* Encode into an ASN.1 sequence. */
+        let vec = vec![asn1_rs::Integer::new(&r), asn1_rs::Integer::new(&s)];
+        let seq = match asn1_rs::Sequence::from_iter_to_der(vec.iter()) {
+            Err(_e) => return Err(BalrogError::SignatureDecodeError),
+            Ok(x) => x,
+        };
+
+        /* Encode it. */
+        seq.to_der_vec()
+            .map_err(|_e| BalrogError::SignatureDecodeError)
+    }
+
+    /* Verify a content signature against the certificate chain. */
+    pub fn verify_content_signature(
+        &self,
+        input: &[u8],
+        signature: &str,
+    ) -> Result<(), BalrogError> {
+        /* To be authorized for code signing, every certificate in the chain needs the
+         * code signing bit in the extended key usage to be set. */
+        for x in self.chain.iter() {
+            let code_sign_ext = match x.extended_key_usage() {
+                Err(_e) => false,
+                Ok(None) => false,
+                Ok(Some(x)) => x.value.code_signing,
+            };
+            if !code_sign_ext {
+                return Err(BalrogError::CodeSignUnauthorized);
+            }
+        }
+
+        let leaf = match self.chain.first() {
+            None => return Err(BalrogError::CertificateNotFound),
+            Some(x) => x,
+        };
+
+        /*  To verify the code signature, we must:
+         *  1. Parse the signature into an asn1_rs::asn1_types::bitstring::Bitstring
+         *  2. Extract the public key from the leaf certificate.
+         *  3. Call x509_parser::verify::verify_signature(
+         *       public_key,
+         *       public_key.algothim,
+         *       signature_bitstring,
+         *       input
+         *     )
+         */
+
+        /* Parse the signature into an ASN1 bitstring. */
+        let sig_decode = match data_encoding::BASE64URL_NOPAD.decode(signature.as_bytes()) {
+            Err(_e) => return Err(BalrogError::SignatureDecodeError),
+            Ok(x) => x,
+        };
+        let sig_der = Self::ecdsa_signature_to_asn1(sig_decode.as_slice())?;
+        let sig_asn1 = asn1_rs::BitString::new(0, sig_der.as_slice());
+        let pubkey = leaf.public_key();
+
+        /*
+         * TODO: Should this be determined by looking at the public key? Or is it
+         * safe to assume an elliptic key with SHA384?
+         *
+         * The rc_crypto crate assumes NIST P-384/SHA384 everywhere.
+         * The golang library chooses a SHA2 that matches the pubkey key size.
+         * Do we care about the RSA/DSA ecosystem?
```

Comment:
If this question was for me... I really do not know. Is the anyone we can ask? If the figuring out is meant to happen as a follow up, please file a ticket :) In that case though, I feel like we should keep the same behaviour as the golang impl.

Reply:
The question was really for myself. I am noting that of all the implementations of this algorithm there is disagreement about which hash algorithm to use when taking the digest of the message data. The golang implementation seems to be the outlier by supporting `SHA256` but the majority of implementations assume `SHA384` As the code stands right now we agree with `rc_crypto` and Firefox's NSS library.

Reply:
The note about RSA/DSA is because we could in theory receive an RSA public key in the leaf certificate and there would still be a logical way to verify the signature, but we ignore that possibility and assume an elliptic curve key everywhere. If this were to occur then we would get a `X509Error::SignatureValidationError`

On the topic of an RSA leaf certificate, we are at least consistent with every other implementation.

Reply:
Ok, well if we are compliant with that then I guess we are fine. It's not like random places are creating this certificate chains for us, correct? So we can be more sure than not that whatever Firefox is doing is the correct.

Either way it looks weird to have this comment in here then, since we are ~kinda~ sure about what we are doing. Can we remove or reword it so sound more sure? 🆘

Reply:
After having a bit more of a ponder about it, I think the golang implementation is the one that's most correct, and it's the one that is generating the signatures for us anyways. So I opted to update the implementation to match.